node/summary-calculation-rpc/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+    }
+
+    fn get_cached_summary(&self, from_block: u32, to_block: u32) -> Option<[u8; 32]> {
+        if let Some(ref storage_mutex) = self.offchain_storage {
+            if let Ok(storage) = storage_mutex.lock() {
+                let key = (from_block, to_block).encode();
+                storage.get(CACHE_PREFIX, &key).and_then(|data| {
+                    if data.len() == 32 {
+                        let mut array = [0u8; 32];
+                        array.copy_from_slice(&data);
+                        Some(array)
+                    } else {
+                        None
+                    }
+                })
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    
+    fn set_cached_summary(&self, from_block: u32, to_block: u32, root: [u8; 32]) {
+        if let Some(ref storage_mutex) = self.offchain_storage {
+            if let Ok(mut storage) = storage_mutex.lock() {
+                let key = (from_block, to_block).encode();
+                storage.set(CACHE_PREFIX, &key, &root.to_vec());
+            }
+        }
+    }
+
+
+}
+
+impl<C, Block> SummaryCalculationProvider<C, Block, ()> {
+    pub fn new_without_storage(client: Arc<C>) -> Self {
+        Self { client, offchain_storage: None, _marker: Default::default() }
+    }
+}
+
+impl<C, Block, O> SummaryCalculationProviderRpcServer for SummaryCalculationProvider<C, Block, O>
+where
+    Block: BlockT,
+    C: Send + Sync + 'static + BlockBackend<Block> + UsageProvider<Block> + HeaderBackend<Block>,
+    O: OffchainStorage + 'static,
+    AccountId: Clone + std::fmt::Display + Codec,
+{
+    fn get_summary_calculation(&self, from_block: u32, to_block: u32) -> RpcResult<String> {
+        let finalized_block_number: u32 = self.client.info().finalized_number.saturated_into();
+
+        if to_block <= finalized_block_number {
+            if let Some(cached_root) = self.get_cached_summary(from_block, to_block) {
+                return Ok(hex::encode(cached_root));
+            }
+        }
+
+        let extrinsics =
+            fetch_extrinsics_from_client::<Block, C>(&self.client, from_block, to_block).map_err(
+                |e| jsonrpsee::core::Error::Custom(format!("Error fetching extrinsics: {:?}", e)),
+            )?;
+
+        let (result, root_bytes) = if !extrinsics.is_empty() {
+            let root_hash = generate_tree_root(extrinsics).map_err(|e| {
+                jsonrpsee::core::Error::Custom(format!("Error generating tree root: {:?}", e))
+            })?;
+
+            let root_bytes = root_hash.0; // Extract [u8; 32] from H256
+            (hex::encode(root_bytes), root_bytes)
+        } else {
+            let empty_root = [0; 32];
+            (hex::encode(empty_root), empty_root)
+        };
+
+        if to_block <= finalized_block_number {
+            self.set_cached_summary(from_block, to_block, root_bytes);
```

Comment:
The cached data should be invalidated after a period and the summary will need to recalculate.